lib/web.ts

```diff
@@ -233,30 +233,32 @@ export default class AutoConsent {
     return cmp
   }
 
-  detectPopups(cmps: AutoCMP[], onFirstPopupAppears: (cmp: AutoCMP) => unknown): Promise<AutoCMP[]> {
-    return new Promise(resolve => {
-      const foundCmps: AutoCMP[] = []
-      let completed = 0
-
-      for (let i = 0, l = cmps.length; i < l; i++) {
-        this.detectPopup(cmps[i])
-          .then(cmp => {
-            foundCmps.push(cmp)
-            
-            if (foundCmps.length === 1) {
-              onFirstPopupAppears(cmp)
-            }
-          })
-          .catch(error => {
-            enableLogs && console.warn(`error waiting for a popup for ${cmps[i].name}`, error)
-          })
-          .finally(() => {
-            if (++completed === l) {
-              resolve(foundCmps)
-            }
-          })
+  async detectPopups(cmps: AutoCMP[], onFirstPopupAppears: (cmp: AutoCMP) => unknown) {
+    const tasks = cmps.map(
+      cmp => this.detectPopup(cmp)
+        // Handle errors immediately and propagate the error to next handler: Promise.allSettled
+        // If we gracefully return errors, Promise.race needs another evaluation
+        .catch(error => {
+          enableLogs && console.warn(`error waiting for a popup for ${cmp.name}`, error)
+
+          throw error
+        })
+    )
+
+    const firstCmpWithPopup = await Promise.race(tasks)
```

Comment:
As far as I can tell, this will throw an error if the first task to settle is a rejection. Perhaps `Promise.any()` could be a better choice here?

Reply:
also, as a side note, I think these tricky places could benefit from [unit tests](https://github.com/duckduckgo/autoconsent/issues/333), but I think we can keep it out of scope of this PR.

Reply:
You’re right. I’m supposed to use .any instead of .race. I’ll fix it as soon as I arrive to my workspace.